modules/xfem/include/userobjects/MeshCut3DUserObject.h

```diff
@@ -0,0 +1,214 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#ifndef MESH_CUT_3D_USEROBJECT_H
+#define MESH_CUT_3D_USEROBJECT_H
+
+#include "GeometricCutUserObject.h"
+
+#include <array>
+
+class MeshCut3DUserObject;
+class Function;
+
+template <>
+InputParameters validParams<MeshCut3DUserObject>();
+
+/**
+ * MeshCut3DUserObject: (1) reads in a mesh describing the crack surface,
+ * (2) uses the mesh to do initial cutting of 3D elements, and
+ * (3) grows the mesh based on prescribed growth functions.
+ */
+
+class MeshCut3DUserObject : public GeometricCutUserObject
```

Comment:
With as much capability as we are adding, eventually it would be nice to migrate this capability to the ElementalUserObject system. There's a lot of time and complexity going into these objects to be put under a GeneralUserObject. None of this code will scale.

Reply:
I have a PR (#10557) that moves all of these classes over to be based on ElementUserObject. That PR breaks a test, and it has been sitting there for a while because we were trying to sort out the bug that was causing it. We almost have a fix for that, so this will happen soon.  This PR and that one will have a conflict, so one of them needs to go in first. I'd like to get this in first, and then get it to play nice with the move to ElementUserObject, which won't be hard.